### PR TITLE
Allow a tsh aws to proxy any command

### PR DESCRIPTION
**Pull request by dballenger**
As a possible solution to #19551, allow a user to specify a custom command to execute instead of just hardcoding "aws" as the backend command to execute.  This should allow for a seamless experience using custom/non-awscli tooling with Teleport vs having to manually launch a proxy and paste env variables.

### CODE REVIEW

**Comment by philip-teleport** [comment]
A full solution to https://github.com/gravitational/teleport/issues/19551 should cater for use cases where external tools/processes (such as Terraform) need to run across multiple AWS accounts, where you would run multiple `tsh proxy aws` processes in parallel with different credentials.
The suggestion in 19551 to "create a profile in the default AWS config file (~/.aws/config) or a deterministic location that we can pass to other tooling via AWS_CONFIG_FILE" would support this use case and is supported in the AWS Terraform Provider which can be configured to use multiple profiles with aliases.

**Comment by dballenger** [comment]
> A full solution to #19551 should cater for use cases where external tools/processes (such as Terraform) need to run across multiple AWS accounts, where you would run multiple `tsh proxy aws` processes in parallel with different credentials.

You can do something similar here, at least when it's one account per invocation (which is the current state of affairs with the `tsh proxy aws` method, isn't it?).  I've been using this across multiple aws accounts for terraform (against a single account at a time) without any issues.

`tsh aws --app prod-aws --aws-command terraform -- plan` and then 
`tsh aws --app staging-aws --aws-command terraform -- plan`, etc.  Is there something else you're referring to do that tsh proxy allows you to do today, that my patch doesn't?

I was the originator of the Issue / request in question (just through our customer support channel), and at the time the solution there was what seemed easiest, until I looked into the actual logic behind how `tsh aws` works.  There's probably room for both solutions here, though.

File: tool/tsh/tsh.go [hunk 1]
```diff
@@ -629,6 +631,7 @@ func Run(ctx context.Context, args []string, opts ...cliOption) error {
 	aws.Flag("app", "Optional Name of the AWS application to use if logged into multiple.").StringVar(&cf.AppName)
 	aws.Flag("endpoint-url", "Run local proxy to serve as an AWS endpoint URL. If not specified, local proxy serves as an HTTPS proxy.").
 		Short('e').Hidden().BoolVar(&cf.AWSEndpointURLMode)
+	aws.Flag("aws-command", "Execute different commands (e.g. terraform) under Teleport credentials").StringVar(&cf.AWSCommand)
```
Comment (r0mant) [inline tool/tsh/tsh.go:634]: I think the flag name should be `--` so it's consistent with how other CLI tools accept extra command arguments.

**Review by r0mant — COMMENTED** [review]
@smallinsky @greedy52 Can you please review this?

**Comment by greedy52** [comment]
@r0mant @smallinsky we had a discussion on this capability when implementing the proxy. 
https://github.com/gravitational/teleport/pull/10025#discussion_r864724233

Should the product team review this new flag? Ideally, we want to apply the same concept to Azure and GCP as well.

**Comment by Jasstkn** [comment]
> > A full solution to #19551 should cater for use cases where external tools/processes (such as Terraform) need to run across multiple AWS accounts, where you would run multiple `tsh proxy aws` processes in parallel with different credentials.
> 
> You can do something similar here, at least when it's one account per invocation (which is the current state of affairs with the `tsh proxy aws` method, isn't it?). I've been using this across multiple aws accounts for terraform (against a single account at a time) without any issues.
> 
> `tsh aws --app prod-aws --aws-command terraform -- plan` and then `tsh aws --app staging-aws --aws-command terraform -- plan`, etc. Is there something else you're referring to do that tsh proxy allows you to do today, that my patch doesn't?
> 
> I was the originator of the Issue / request in question (just through our customer support channel), and at the time the solution there was what seemed easiest, until I looked into the actual logic behind how `tsh aws` works. There's probably room for both solutions here, though.

That will work without trouble for sure but what about the case when it's several AWS accounts(=aws profiles) that are being used in the terraform. It can be the case for reading data, state and etc.

**Review by smallinsky — COMMENTED** [review]
I think that this will be usefully functionally. Thanks for the contribution. 
I have left one regarding UX flag naming. 

Regarding:
> A full solution to https://github.com/gravitational/teleport/issues/19551 should cater for use cases where external tools/processes (such as Terraform) need to run across multiple AWS accounts, where you would run multiple tsh proxy aws processes in parallel with different credentials.

For that logic of teleport AWS local proxy needs to be aligned. Right now the local proxy can only handle only one teleport aws app.

File: tool/tsh/tsh.go (hunk 1 above)
Comment (greedy52) [reply tool/tsh/tsh.go:634]: Personally, I would keep `tsh aws` as it is. Here are my reasonings:
- `tsh aws/az/gcloud` are designed to be simple replacements for `aws/az/gcloud` command.
- I would call it a "coincident" that the setup for AWS CLI works for `terraform`. Many applications require extra setup (e.g.  `NODE_EXTRA_CA_CERTS=<ca-path>` for node, `AWS_SDK_LOAD_CONFIG=true` for golang SDK v2, trust store for Java)

I would propose to add this functionality to the proxy command, e.g.:
```
tsh proxy aws -- terraform init
```
The proxy command is more flexible with `-p` or `-f` flat etc. And possibly we can do cool things like these in the future.
```
tsh proxy aws --format nodejs -- my-node-program
tsh proxy aws --app aws-account-1 --app aws-acount-2 -- terraform init
```

If we agree on `tsh proxy aws -- <command and args>`, I can make a draft PR for that.

If we prefer to keep it in `tsh aws`, I personally like `--exec` better than `--aws-command`. Though I may prefer `--exec` to be a boolean. I don't think we can use `--` as a flag name directly. `--` is built-in with special logic for separating flags from positional arguments.

Comment (dballenger) [reply tool/tsh/tsh.go:634]: > Can we rename this to `--exec` flag. I think that duplicate the aws prefix like `tsh aws app1 --aws-comand ...` so the UI will look as follow: `tsh aws app1 --exec ...`
> 
> Also we have `tsh az` `tsh gcp` command and the `--exec` flag seems to be more generic. @greedy52 wdyt ?

Yeah, I can push this change, I think exec is a lot better than aws-command as a flag, even not considering azure and gcp functionality you guys have.

**Review by smallinsky — APPROVED** [review]
@greedy52 Could you also take a second look  ?

File: tool/tsh/tsh.go [hunk 2]
```diff
@@ -336,6 +336,8 @@ type CLIConf struct {
 	// LocalProxyTunnel specifies whether local proxy will open auth'd tunnel.
 	LocalProxyTunnel bool
 
+	// Command to run via tsh aws
+	Exec string
```
Comment (greedy52) [inline tool/tsh/tsh.go:340]: nit
```suggestion
	// Exec is the command to run via tsh aws.
	Exec string
```

**Review by greedy52 — APPROVED** [review]

File: tool/tsh/tsh.go (hunk 2 above)
Comment (dballenger) [reply tool/tsh/tsh.go]: just pushed that change up @greedy52

**Review by r0mant — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

**Comment by othmane399** [comment]
Hello @smallinsky  @greedy52 thanks for this pull request, but this doesn't solve presigning requests issues.

Basically running a command like this one 
`tsh aws s3 presign s3://DOC-EXAMPLE-BUCKET/test2.txt` 
will generate a wrong signature like this one as it not use the real credentials
```
https://s3.amazonaws.com/DOC-EXAMPLE-BUCKET/test2.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=1e23f5f3301b9655ec97%2F20230321%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230321T225914Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=c03e19034446ce141941f0b343e6853064a4590cbcc2c6dc36badd25ece2c17c
```
And this is because the [signing algorithm](https://github.com/aws/aws-sdk-go/blob/28b1fd28879898d7330b3765103169c3d2900847/aws/request/request.go#L331) is only done locally and doesn't interact with the AWS API, so the signing logic should be done at the proxy level IMO.

Is there a solution to fix that ?

Cheers

**Comment by greedy52** [comment]
@othmane399 thanks for sharing this. I will create issue #23435 to track this (and some workarounds).

**Comment by dballenger** [comment]
@greedy52 thanks for the review and merging into master here, when might this be backported into the v12 branch (which I think brew is tracking right now)?

**Comment by greedy52** [comment]
@dballenger it's now released in v12.1.5

https://github.com/gravitational/teleport/blob/v12.1.5/tool/tsh/tsh.go#L666

cheers

**Comment by thameezb** [comment]
> > > A full solution to #19551 should cater for use cases where external tools/processes (such as Terraform) need to run across multiple AWS accounts, where you would run multiple `tsh proxy aws` processes in parallel with different credentials.
> > 
> > 
> > You can do something similar here, at least when it's one account per invocation (which is the current state of affairs with the `tsh proxy aws` method, isn't it?). I've been using this across multiple aws accounts for terraform (against a single account at a time) without any issues.
> > `tsh aws --app prod-aws --aws-command terraform -- plan` and then `tsh aws --app staging-aws --aws-command terraform -- plan`, etc. Is there something else you're referring to do that tsh proxy allows you to do today, that my patch doesn't?
> > I was the originator of the Issue / request in question (just through our customer support channel), and at the time the solution there was what seemed easiest, until I looked into the actual logic behind how `tsh aws` works. There's probably room for both solutions here, though.
> 
> That will work without trouble for sure but what about the case when it's several AWS accounts(=aws profiles) that are being used in the terraform. It can be the case for reading data, state and etc.

Any movement on support for this requirement?

**Comment by greedy52** [comment]
> > > > A full solution to #19551 should cater for use cases where external tools/processes (such as Terraform) need to run across multiple AWS accounts, where you would run multiple `tsh proxy aws` processes in parallel with different credentials.
> > > 
> > > 
> > > You can do something similar here, at least when it's one account per invocation (which is the current state of affairs with the `tsh proxy aws` method, isn't it?). I've been using this across multiple aws accounts for terraform (against a single account at a time) without any issues.
> > > `tsh aws --app prod-aws --aws-command terraform -- plan` and then `tsh aws --app staging-aws --aws-command terraform -- plan`, etc. Is there something else you're referring to do that tsh proxy allows you to do today, that my patch doesn't?
> > > I was the originator of the Issue / request in question (just through our customer support channel), and at the time the solution there was what seemed easiest, until I looked into the actual logic behind how `tsh aws` works. There's probably room for both solutions here, though.
> > 
> > 
> > That will work without trouble for sure but what about the case when it's several AWS accounts(=aws profiles) that are being used in the terraform. It can be the case for reading data, state and etc.
> 
> Any movement on support for this requirement?

Not at the moment. I think the easiest workaround would be to have an all-in-one role (that accessed by Teleport) to assume roles in other accounts and use `assume_role` in the AWS provider block. Of course, these assumed sessions will bypass Teleport.

[AWS roles anywhere](https://docs.aws.amazon.com/rolesanywhere/latest/userguide/introduction.html) is another potential workaround. We have a guide on using [Teleport's workload identity with AWS roles anywhere](https://goteleport.com/docs/enroll-resources/machine-id/workload-identity/aws-roles-anywhere/). Alternatively, you may be able to use the user certificate your `tsh login` gives you which is signed by the user CA (`tctl auth export --type user`). Will require a lot of set up though.

**Comment by thameezb** [comment]
> > > > > A full solution to #19551 should cater for use cases where external tools/processes (such as Terraform) need to run across multiple AWS accounts, where you would run multiple `tsh proxy aws` processes in parallel with different credentials.
> > > > 
> > > > 
> > > > You can do something similar here, at least when it's one account per invocation (which is the current state of affairs with the `tsh proxy aws` method, isn't it?). I've been using this across multiple aws accounts for terraform (against a single account at a time) without any issues.
> > > > `tsh aws --app prod-aws --aws-command terraform -- plan` and then `tsh aws --app staging-aws --aws-command terraform -- plan`, etc. Is there something else you're referring to do that tsh proxy allows you to do today, that my patch doesn't?
> > > > I was the originator of the Issue / request in question (just through our customer support channel), and at the time the solution there was what seemed easiest, until I looked into the actual logic behind how `tsh aws` works. There's probably room for both solutions here, though.
> > > 
> > > 
> > > That will work without trouble for sure but what about the case when it's several AWS accounts(=aws profiles) that are being used in the terraform. It can be the case for reading data, state and etc.
> > 
> > 
> > Any movement on support for this requirement?
> 
> Not at the moment. I think the easiest workaround would be to have an all-in-one role (that accessed by Teleport) to assume roles in other accounts and use `assume_role` in the AWS provider block.
> 
> [AWS roles anywhere](https://docs.aws.amazon.com/rolesanywhere/latest/userguide/introduction.html) is another potential workaround. We have a guide on using [Teleport's workload identity with AWS roles anywhere](https://goteleport.com/docs/enroll-resources/machine-id/workload-identity/aws-roles-anywhere/). Alternatively, you may be able to use the user certificate your `tsh login` gives you which is signed by the user CA (`tctl auth export --type user`). Will require a lot of set up though.

>  all-in-one role (that accessed by Teleport) to assume roles in other accounts and use assume_role in the AWS provider block.

This is quite a security issue, as if that single role is compromised, one would have practically admin powers in all subsequent accounts. 
It also removes the usefulness of using teleport to ensure that only team A can access account A and B, while team C can access accounts B and C

> [AWS roles anywhere](https://docs.aws.amazon.com/rolesanywhere/latest/userguide/introduction.html) is another potential workaround. We have a guide on using [Teleport's workload identity with AWS roles anywhere](https://goteleport.com/docs/enroll-resources/machine-id/workload-identity/aws-roles-anywhere/).

This can work, but requires quite a bit more config, and removes a benefit of teleport (in this case the audit log).

This issue has been cooking for about a year now (and the requirement to use >1 AWS profile in terraform is quite common). Surely there can be a more elegant native-teleport manner to achieve this goal? Which maintains a single role per account, the RBAC aspects of teleport, and the audit trail (using the `external_credential` provider in the AWS CLI seems like a perfect use for this)

**Comment by thameezb** [comment]
BTW @greedy52 docs are slightly out of date for [Teleport's workload identity with AWS roles anywhere](https://goteleport.com/docs/enroll-resources/machine-id/workload-identity/aws-roles-anywhere/).

` tctl auth export --type spiffe` should read ` tctl auth export --type tls-spiffe`

**Pull request merged**